test/pipelinerun_test.go

```diff
@@ -33,12 +34,102 @@ import (
 	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
 )
 
+// dumpAndTearDown will, when a test fails, print out all CRD objects as YAML
+// in the relevant namespace before tearing the namespace down. It will also
+// get the logs of all pods in the namespace. If the tests do not fail then the
+// namespace is simply torn down.
+func dumpAndTearDown(t *testing.T, logger *logging.BaseLogger, cs *clients, namespace string) {
+	defer tearDown(logger, cs.KubeClient, namespace)
```

Comment:
this is a pretty personal preference, but i'd rather see the `tearDown` function call `dump` on failure, e.g.:

in the test:
```
	knativetest.CleanupOnInterrupt(func() { tearDown(logger, c.KubeClient, namespace) }, logger)
	defer tearDown(logger, c.KubeClient, namespace)
```

in `tearDown`:
```
if t.Failed() {
		dumpResources(...)
	}
... other teardown stuff ...
```

Reply:
and once we're happy with this function we could move it somewhere like `init_test.go`